File: src/features/estimates/dto/update-estimate.dto.ts
import { Transform, Type } from 'class-transformer';
import { IsArray, IsIn, IsOptional, IsString, ValidateNested } from 'class-validator';
import { EstimateLineItemDto } from './estimate-line-item.dto';

export class UpdateEstimateDto {
  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  name?: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  description?: string;

  @IsOptional()
  @IsString()
  @Transform(({ value }) => (typeof value === 'string' ? value.trim() : value))
  notes?: string;

  @IsOptional()
  @IsIn(['draft', 'final'])
  status?: 'draft' | 'final';

  @IsOptional()
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => EstimateLineItemDto)
  lineItems?: EstimateLineItemDto[];
}

File: spring-boot-api/src/main/java/com/mytegroup/api/dto/estimates/UpdateEstimateDto.java
package com.mytegroup.api.dto.estimates;

import com.mytegroup.api.entity.enums.projects.EstimateStatus;
import jakarta.validation.Valid;

import java.util.List;

public record UpdateEstimateDto(
    String name,
    String description,
    String notes,
    EstimateStatus status,
    @Valid
    List<EstimateLineItemDto> lineItems
) {
    public UpdateEstimateDto {
        if (name != null) {
            name = name.trim();
        }
        if (description != null) {
            description = description.trim();
        }
        if (notes != null) {
            notes = notes.trim();
        }
    }
}

